File: applications/rtkvarianobigeometry/rtkvarianobigeometry.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkvarianobigeometry_ggo.h"
#include "rtkGgoFunctions.h"

#include "rtkVarianObiGeometryReader.h"
#include "rtkThreeDCircularProjectionGeometryXMLFile.h"

#include <itkRegularExpressionSeriesFileNames.h>

int main(int argc, char * argv[])
{
  GGO(rtkvarianobigeometry, args_info);

  // Generate file names of projections
  itk::RegularExpressionSeriesFileNames::Pointer names = itk::RegularExpressionSeriesFileNames::New();
  names->SetDirectory(args_info.path_arg);
  names->SetNumericSort(false);
  names->SetRegularExpression(args_info.regexp_arg);
  names->SetSubMatch(0);

  // Create geometry reader
  rtk::VarianObiGeometryReader::Pointer reader;
  reader = rtk::VarianObiGeometryReader::New();
  reader->SetXMLFileName(args_info.xml_file_arg);
  reader->SetProjectionsFileNames(names->GetFileNames());
  TRY_AND_EXIT_ON_ITK_EXCEPTION( reader->UpdateOutputData() );

  // Write
  rtk::ThreeDCircularProjectionGeometryXMLFileWriter::Pointer xmlWriter =
    rtk::ThreeDCircularProjectionGeometryXMLFileWriter::New();
  xmlWriter->SetFilename(args_info.output_arg);
  xmlWriter->SetObject( reader->GetGeometry() );
  TRY_AND_EXIT_ON_ITK_EXCEPTION( xmlWriter->WriteFile() )

  return EXIT_SUCCESS;
}

File: applications/rtkvarianobigeometry/rtkvarianobigeometry.py
#!/usr/bin/env python
import argparse
import itk
from itk import RTK as rtk


def build_parser():
    # Argument parsing
    parser = rtk.RTKArgumentParser(
        description="Creates an RTK geometry file from a Varian OBI acquisition."
    )

    parser.add_argument(
        "--xml_file",
        "-x",
        help="Varian OBI XML information file on projections",
        required=True,
    )
    parser.add_argument("--output", "-o", help="Output file name", required=True)
    parser.add_argument(
        "--path", "-p", help="Path containing projections", required=True
    )
    parser.add_argument(
        "--regexp",
        "-r",
        help="Regular expression to select projection files in path",
        required=True,
    )

    # Parse the command line arguments
    return parser


def process(args: argparse.Namespace):

    names = itk.RegularExpressionSeriesFileNames.New()
    names.SetDirectory(args.path)
    names.SetRegularExpression(args.regexp)

    reader = rtk.VarianObiGeometryReader.New()
    reader.SetXMLFileName(args.xml_file)
    reader.SetProjectionsFileNames(names.GetFileNames())
    reader.UpdateOutputData()

    rtk.write_geometry(reader.GetGeometry(), args.output)


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
